File: agentprimordia/internal/governance/policy.go
// Package governance 实现"策略即代码"：用 YAML 定义 Agent 行为策略，
// 运行时通过 PolicyEnforcer 强制执行（工具限制 / 成本限制 / 输出护栏 / 行为约束）。
//
// 设计取舍：
//   - 策略以 YAML 描述，加载期解析为 Policy 结构；
//   - 运行时通过 Enforcer 接口暴露检查能力，供上层（agent 包）依赖，
//     避免 governance 反向引用 agent（遵循模块依赖方向）；
//   - 并发安全：PolicyEnforcer 用 sync.Mutex 保护运行时计数与累计成本。
package governance

import (
	"errors"
	"regexp"
)

// 策略执行相关错误
var (
	// ErrToolCallLimitExceeded 单次运行内某工具调用次数/会话总调用超过上限
	ErrToolCallLimitExceeded = errors.New("tool call limit exceeded")
	// ErrCostLimitExceeded 成本超过策略上限
	ErrCostLimitExceeded = errors.New("cost limit exceeded")
	// ErrBlockedArgument 工具参数命中禁止模式
	ErrBlockedArgument = errors.New("blocked argument pattern")
	// ErrOutputTooLong 输出长度超过策略上限
	ErrOutputTooLong = errors.New("output exceeds max length")
	// ErrPIIDetected 输出触发 strict PII 过滤
	ErrPIIDetected = errors.New("output triggered strict PII filter")
	// ErrPolicyNotFound 未加载策略
	ErrPolicyNotFound = errors.New("policy not loaded")
)

// 常见 PII 模式（strict 模式下的简化检测）
var (
	rePhone = regexp.MustCompile(`1[3-9]\d{9}`)
	reID    = regexp.MustCompile(`\d{17}[\dXx]`)
)

// Policy 顶层策略定义
type Policy struct {
	APIVersion string         `yaml:"apiVersion"`
	Kind       string         `yaml:"kind"`
	Metadata   PolicyMetadata `yaml:"metadata"`
	Spec       PolicySpec     `yaml:"spec"`
}

// PolicyMetadata 策略元信息
type PolicyMetadata struct {
	Name string `yaml:"name"`
}

// PolicySpec 策略规约
type PolicySpec struct {
	ToolRestrictions    []ToolRestriction   `yaml:"toolRestrictions"`
	CostLimits          CostLimits          `yaml:"costLimits"`
	OutputGuardrail     OutputGuardrail     `yaml:"outputGuardrail"`
	BehaviorConstraints BehaviorConstraints `yaml:"behaviorConstraints"`
}

// ToolRestriction 单工具限制
type ToolRestriction struct {
	Tool            string   `yaml:"tool"`
	RequireApproval bool     `yaml:"requireApproval"`
	MaxCallsPerRun  int      `yaml:"maxCallsPerRun"`
	BlockedArgs     []string `yaml:"blockedArgs"`
	AllowedDomains  []string `yaml:"allowedDomains"`
}

// CostLimits 成本限制（美元）
type CostLimits struct {
	PerRequest float64 `yaml:"perRequest"`
	PerDay     float64 `yaml:"perDay"`
	PerSession float64 `yaml:"perSession"`
}

// OutputGuardrail 输出护栏
type OutputGuardrail struct {
	PIIFilter      string `yaml:"piiFilter"`      // strict | moderate | off
	InjectionBlock bool   `yaml:"injectionBlock"` // 是否拦截注入
	MaxLength      int    `yaml:"maxLength"`
}

// BehaviorConstraints 行为约束
type BehaviorConstraints struct {
	MaxTurns          int  `yaml:"maxTurns"`
	MaxToolCalls      int  `yaml:"maxToolCalls"`
	RequireReflection bool `yaml:"requireReflection"`
}

File: sdk/typescript/src/governance/policy.ts
/**
 * governance/policy.ts — 策略执行器
 *
 * 对齐 Go internal/governance/policy_enforcer.go 的核心能力。
 * Stability: Stable
 */

import type { Policy, EnforcerSnapshot, AuditEvent, AuditEventType } from './types.js';

/** 策略执行错误 */
export class PolicyViolationError extends Error {
  readonly code: string;
  constructor(code: string, message: string) {
    super(message);
    this.name = 'PolicyViolationError';
    this.code = code;
  }
}

/** 审计日志接口 */
export interface AuditLogger {
  log(event: AuditEvent): void;
  close(): void;
}

/** 内存审计日志（测试/轻量场景） */
export class InMemoryAuditLogger implements AuditLogger {
  events: AuditEvent[] = [];

  log(event: AuditEvent): void {
    this.events.push(event);
  }

  close(): void {
    // no-op
  }
}

let auditIdCounter = 0;

/**
 * 策略执行器（运行时强制）。
 *
 * 检查工具调用、成本和输出是否符合策略定义。
 * 线程安全：JS 单线程模型下天然安全。
 */
export class PolicyEnforcer {
  private policy: Policy;
  private toolCallCount = new Map<string, number>();
  private totalCost = 0;
  private totalToolCalls = 0;
  private auditLog?: AuditLogger;
  private agentId: string;

  constructor(policy: Policy, opts?: { auditLog?: AuditLogger; agentId?: string }) {
    this.policy = policy;
    this.auditLog = opts?.auditLog;
    this.agentId = opts?.agentId ?? 'unknown';
  }

  /** 检查工具调用是否允许 */
  checkToolCall(toolName: string, args: string): void {
    const restrictions = this.policy.spec.toolRestrictions ?? [];
    const restriction = restrictions.find((r) => r.tool === toolName);

    if (restriction) {
      // 检查调用次数上限
      if (restriction.maxCallsPerRun) {
        const count = this.toolCallCount.get(toolName) ?? 0;
        if (count >= restriction.maxCallsPerRun) {
          this.emitAudit('tool_call_blocked', toolName, `tool ${toolName} exceeded max calls per run`);
          throw new PolicyViolationError('TOOL_CALL_LIMIT', `Tool ${toolName} call limit exceeded`);
        }
      }

      // 检查禁止参数模式
      if (restriction.blockedArgs) {
        for (const pattern of restriction.blockedArgs) {
          if (args.includes(pattern)) {
            this.emitAudit('tool_call_blocked', toolName, `blocked argument pattern: ${pattern}`);
            throw new PolicyViolationError('BLOCKED_ARGUMENT', `Blocked argument pattern: ${pattern}`);
          }
        }
      }
    }

    // 全局工具调用上限
    const maxToolCalls = this.policy.spec.behaviorConstraints?.maxToolCalls;
    if (maxToolCalls && this.totalToolCalls >= maxToolCalls) {
      this.emitAudit('tool_call_blocked', toolName, 'global tool call limit exceeded');
      throw new PolicyViolationError('TOOL_CALL_LIMIT', 'Global tool call limit exceeded');
    }
  }

  /** 检查成本是否允许 */
  checkCost(cost: number): void {
    const limits = this.policy.spec.costLimits;
    if (!limits) return;

    if (limits.perRequest && cost > limits.perRequest) {
      this.emitAudit('cost_exceeded', undefined, `per-request cost ${cost} exceeds ${limits.perRequest}`);
      throw new PolicyViolationError('COST_LIMIT', `Per-request cost limit exceeded`);
    }

    if (limits.perDay && this.totalCost + cost > limits.perDay) {
      this.emitAudit('cost_exceeded', undefined, `daily cost would exceed ${limits.perDay}`);
      throw new PolicyViolationError('COST_LIMIT', `Daily cost limit exceeded`);
    }
  }

  /** 检查输出是否合规 */
  checkOutput(output: string): void {
    const guardrail = this.policy.spec.outputGuardrail;
    if (!guardrail) return;

    // 输出长度检查
    if (guardrail.maxLength && output.length > guardrail.maxLength) {
      this.emitAudit('output_blocked', undefined, `output length ${output.length} exceeds ${guardrail.maxLength}`);
      throw new PolicyViolationError('OUTPUT_TOO_LONG', `Output exceeds max length`);
    }

    // PII 检测（strict 模式）
    if (guardrail.piiFilter === 'strict') {
      const phoneRe = /1[3-9]\d{9}/;
      const idRe = /\d{17}[\dXx]/;
      if (phoneRe.test(output) || idRe.test(output)) {
        this.emitAudit('pii_detected', undefined, 'strict PII filter triggered');
        throw new PolicyViolationError('PII_DETECTED', `Output triggered strict PII filter`);
      }
    }
  }

  /** 记录工具调用（计数） */
  recordToolCall(toolName: string): void {
    this.toolCallCount.set(toolName, (this.toolCallCount.get(toolName) ?? 0) + 1);
    this.totalToolCalls++;
  }

  /** 记录成本 */
  recordCost(cost: number): void {
    this.totalCost += cost;
  }

  /** 获取运行时快照 */
  snapshot(): EnforcerSnapshot {
    const toolCalls: Record<string, number> = {};
    for (const [k, v] of this.toolCallCount) toolCalls[k] = v;
    return { toolCalls, totalCost: this.totalCost, totalToolCalls: this.totalToolCalls };
  }

  /** 热替换策略 */
  hotSwap(newPolicy: Policy): void {
    this.policy = newPolicy;
    this.emitAudit('policy_hot_swapped', undefined, `policy swapped to ${newPolicy.metadata.name}`);
  }

  private emitAudit(type: AuditEventType, toolName: string | undefined, reason: string): void {
    if (!this.auditLog) return;
    this.auditLog.log({
      id: `audit-${++auditIdCounter}-${Date.now()}`,
      timestamp: new Date().toISOString(),
      type,
      agentId: this.agentId,
      toolName,
      reason,
      severity: type === 'pii_detected' ? 'critical' : 'warning',
    });
  }
}
